src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -0,0 +1,364 @@
+                }
+
+                return pEnvelopedEncodeInfo;
+            }
+
+            //
+            // This returns an allocated native memory block. Its lifetime (and that of any allocated subblocks it may point to) is the shorter that of "hb". 
+            //
+            private static CMSG_RECIPIENT_ENCODE_INFO EncodeRecipientInfo(CmsRecipient recipient, AlgorithmIdentifier contentEncryptionAlgorithm, HeapBlockRetainer hb)
+            {
+                CMsgCmsRecipientChoice recipientChoice;
+                string keyAlgorithmOid = recipient.Certificate.GetKeyAlgorithm();
+                AlgId algId = keyAlgorithmOid.ToAlgId();
+                switch (algId)
+                {
+                    case AlgId.CALG_RSA_KEYX:
+                        recipientChoice = CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT;
+                        break;
+
+                    case AlgId.CALG_DH_SF:
+                    case AlgId.CALG_DH_EPHEM:
+                        recipientChoice = CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT;
+                        break;
+
+                    default:
+                        throw ErrorCode.CRYPT_E_UNKNOWN_ALGO.ToCryptographicException();
+                }
+
+                CMSG_RECIPIENT_ENCODE_INFO recipientEncodeInfo;
+                recipientEncodeInfo.dwRecipientChoice = recipientChoice;
+
+                unsafe
+                {
+                    switch (recipientChoice)
+                    {
+                        case CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT:
+                            recipientEncodeInfo.pCmsRecipientEncodeInfo = (IntPtr)EncodeKeyTransRecipientInfo(recipient, hb);
+                            break;
+
+                        case CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT:
+                            recipientEncodeInfo.pCmsRecipientEncodeInfo = (IntPtr)EncodeKeyAgreeRecipientInfo(recipient, contentEncryptionAlgorithm, hb);
+                            break;
+
+                        default:
+                            throw ErrorCode.CRYPT_E_UNKNOWN_ALGO.ToCryptographicException();
+                    }
+                }
+                return recipientEncodeInfo;
+            }
+
+            //
+            // This returns an allocated native memory block. Its lifetime (and that of any allocated subblocks it may point to) is that of "hb". 
+            //
+            private static unsafe CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO* EncodeKeyTransRecipientInfo(CmsRecipient recipient, HeapBlockRetainer hb)
+            {
+                // "recipient" is a deep-cloned CmsRecipient object whose lifetime this class controls. Because of this, we can pull out the CERT_CONTEXT* and CERT_INFO* pointers without
+                // bringing in all the SafeCertContextHandle machinery, and embed pointers to them in the memory block we return. Yes, this code is scary.
```

Comment:
That comment was written before I found I had to create a SafeCertContextHandle for other parts of the code.